hasktorch/src/Torch/Typed/NN.hs

```diff
@@ -39,24 +41,38 @@ import           Torch.Typed.Factories
 import           Torch.Typed.Functional
 import           Torch.Typed.Tensor
 import           Torch.Typed.Parameter
-
-
-data LinearSpec (inputFeatures :: Nat) (outputFeatures :: Nat)
-                (dtype :: D.DType)
-                (device :: (D.DeviceType, Nat))
+import           Torch.Typed.Device
+
+class HasForward f a b | f a -> b where
+  forward :: f -> a -> b
+  forwardStoch :: f -> a -> IO b
+  forwardStoch = (pure .) . forward
```

Comment:
When we can only define forwardStoch, do we define forward = undefined?
This does not have to be changed.

Reply:
I'm trying something new here. not sure this will stay like this.
I think that `forward` should always be effect-free, while `forwardStoch` can have effects.
I don't believe there are any models people use that are effectful once the randomness is turned off. Therefore, `forward` can probably be always defined.
For instance, note how I defined a `HasForward` instance for `Dropout`:
```Haskell
instance HasForward Dropout (Tensor device dtype shape) (Tensor device dtype shape) where
  forward dropout input = unsafePerformIO $ Torch.Typed.NN.dropout dropout False input
  forwardStoch dropout input = Torch.Typed.NN.dropout dropout True input
```

Reply:
Since randomness isn't the only possible effect (though maybe the most common one), what about forwardIO or forwardM or some other name that reflects the generality?

Reply:
Reminds me, this comes up in VAEs too.

Reply:
`forwardEffectfully`?